torch/csrc/autograd/python_variable_indexing.h

```diff
@@ -1,11 +1,94 @@
 #pragma once
 
+#include <c10/core/SymInt.h>
+#include <c10/util/Exception.h>
+#include <pybind11/detail/common.h>
+#include <pybind11/pytypes.h>
+#include <pyport.h>
+#include <sys/types.h>
 #include <torch/csrc/autograd/python_variable.h>
 #include <torch/csrc/python_headers.h>
+#include <torch/csrc/utils/pybind.h>
+#include <torch/csrc/utils/python_symnode.h>
 
 namespace torch {
 namespace autograd {
 
+struct UnpackedSlice {
+  c10::SymInt start;
+  c10::SymInt stop;
+  c10::SymInt step;
+};
+
+// This mirrors Cpython's PySlice_Unpack method
+inline UnpackedSlice __PySlice_Unpack(PyObject* _r) {
+  PySliceObject* r = (PySliceObject*)_r;
+  /* this is harder to get right than you might think */
+
+  // Py_BUILD_ASSERT replaced because it is not available in all versions
+  static_assert(PY_SSIZE_T_MIN + 1 <= -PY_SSIZE_T_MAX, "Build failed");
```

Comment:
We should use our INDEX_MIN/MAX here instead of the python ssize macros.

Reply:
Why does this assert exist? Do we actually need it?

Reply:
Just to match the CPython code I'm sure.
But this is most likely because we do assume that these values have this property below and so, since we can do this check for free, we make sure it is true.

Reply:
I think we should either keep this or add a comment next to our definition of INDEX_MIN/MAX that we rely on this property